File: app/src/main/kt/estyle/teabaike/activity/CollectionActivity.java
package estyle.teabaike.activity;

import android.content.Context;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.graphics.Color;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;

import java.util.List;
import java.util.Locale;

import javax.inject.Inject;

import estyle.teabaike.R;
import estyle.teabaike.adapter.CollectionAdapter;
import estyle.teabaike.application.TeaBaikeApplication;
import estyle.teabaike.bean.ContentDataBean;
import estyle.teabaike.bean.eventbus.CheckAllCollectionsEvent;
import estyle.teabaike.databinding.ActivityCollectionBinding;
import estyle.teabaike.manager.GreenDaoManager;

public class CollectionActivity extends BaseActivity implements
        CollectionAdapter.OnItemClickListener, CollectionAdapter.OnItemLongClickListener {

    private ActivityCollectionBinding binding;

    private CollectionAdapter mAdapter;

    // 删除功能是否可用
    private boolean mIsDeleteEnabled;

    private Snackbar mSnackbar;

    @Inject
    GreenDaoManager mDBProvider;

    public static void startActivity(Context context) {
        Intent intent = new Intent(context, CollectionActivity.class);
        context.startActivity(intent);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        TeaBaikeApplication.getInstance().getTeaBaikeComponent().inject(this);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_collection);
        EventBus.getDefault().register(this);

        initView();
        initData();
    }

    private void initView() {
        setSupportActionBar(binding.toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        mAdapter = new CollectionAdapter(this);
        mAdapter.setEmptyView(binding.emptyView);
        mAdapter.setOnItemClickListener(this);
        mAdapter.setOnItemLongClickListener(this);
        binding.setAdapter(mAdapter);
    }

    private void initData() {
        List<ContentDataBean> collectionList = mDBProvider.queryCollectionDatas();
        mAdapter.addDatas(collectionList);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                if (mIsDeleteEnabled) {
                    setDeleteEnabled(false);
                } else {
                    finish();
                }
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onItemClick(int position) {
        if (!mIsDeleteEnabled) {
            ContentActivity.startActivity(this, mAdapter.getItemId(position), false);
        } else {
            mAdapter.invertItemStateAtPosition(position);
        }
    }

    @Override
    public void onItemLongClick(int position) {
        if (!mIsDeleteEnabled) {
            mAdapter.invertItemStateAtPosition(position);
            setDeleteEnabled(true);
        }
    }

    // 全选/取消全选按钮
    public void checkAll(View view) {
        TextView checkAllTextView = (TextView) view;
        switch (checkAllTextView.getText().toString()) {
            case "全选":
                mAdapter.setIsCheckedAllItem(true);
                break;
            case "取消":
                mAdapter.setIsCheckedAllItem(false);
                break;
        }
    }

    // 删除选中数据
    public void deleteItem(View view) {
        int deleteCount = mAdapter.deleteCheckedItem();
        String tip = String.format(Locale.getDefault(), getString(R.string.delete_successful), deleteCount);
        setDeleteEnabled(false);
        if (mSnackbar == null) {
            mSnackbar = Snackbar.make(binding.getRoot(), tip, Snackbar.LENGTH_LONG)
                    .setAction(R.string.revoke, v -> mAdapter.restoreTempItem())
                    .setActionTextColor(Color.BLACK)
                    .addCallback(snackBarCallback);
            mSnackbar.getView().setBackgroundResource(R.color.colorAccent);
        } else {
            mSnackbar.setText(tip);
        }
        mSnackbar.show();
    }

    // 设置是否可删除
    private void setDeleteEnabled(boolean isDeleteEnabled) {
        this.mIsDeleteEnabled = isDeleteEnabled;
        mAdapter.setDeleteBoxVisibility(isDeleteEnabled);
        int visibility = isDeleteEnabled ? View.VISIBLE : View.INVISIBLE;
        binding.deleteBtn.setVisibility(visibility);
        binding.checkAllTextView.setVisibility(visibility);
    }

    @Override
    public void onBackPressed() {
        if (mIsDeleteEnabled) {
            setDeleteEnabled(false);
        } else {
            super.onBackPressed();
        }
    }

    // 控制全选按钮文字
    @Subscribe
    public void setCheckAllText(CheckAllCollectionsEvent event) {
        binding.checkAllTextView.setText(event.getCheckAllText());
    }


    // Snackbar消失回调
    private Snackbar.Callback snackBarCallback = new Snackbar.Callback() {
        @Override
        public void onDismissed(Snackbar transientBottomBar, int event) {
            super.onDismissed(transientBottomBar, event);
            mAdapter.deleteData();
        }
    };

    @Override
    protected void onDestroy() {
        super.onDestroy();
        EventBus.getDefault().unregister(this);
        if (mSnackbar != null) {
            mSnackbar.removeCallback(snackBarCallback);
        }
    }

}

File: app/src/main/kt/estyle/teabaike/activity/CollectionActivity.kt
package estyle.teabaike.activity

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import androidx.lifecycle.ViewModelProviders
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import androidx.paging.PagedList
import estyle.base.BaseActivity
import estyle.base.rxjava.DisposableConverter
import estyle.base.rxjava.observer.RefreshObserver
import estyle.base.rxjava.observer.SnackbarObserver
import estyle.base.widget.Snackbar
import estyle.teabaike.R
import estyle.teabaike.adapter.CollectionAdapter
import estyle.teabaike.entity.ContentEntity
import estyle.teabaike.fragment.dialog.DeleteCollectionDialogFragment
import estyle.teabaike.viewmodel.CollectionViewModel
import kotlinx.android.synthetic.main.activity_collection.*

class CollectionActivity : BaseActivity() {

    private val viewModel by lazy { ViewModelProviders.of(this)[CollectionViewModel::class.java] }

    private val adapter by lazy { CollectionAdapter() }

    private val deleteReceiver by lazy { DeleteReceiver() }

    // 删除功能是否可用
    private var mIsDeleteEnabled: Boolean = false

    private val deleteDialog by lazy {
        DeleteCollectionDialogFragment.newInstance().apply {
            positiveCallback = { deleteItems() }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_collection)
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        adapter.emptyView = empty_view
        adapter.onItemClickListener = { position, id ->
            if (!mIsDeleteEnabled) {
                ContentActivity.startActivity(this, id)
            } else {
                adapter.invertItemStateAtPosition(position)
            }
        }
        adapter.onItemLongClickListener = { position, id ->
            if (!mIsDeleteEnabled) {
                adapter.invertItemStateAtPosition(position)
                setDeleteEnabled(true)
            }
        }
        recycler_view.adapter = adapter

        check_all_text_view.setOnClickListener {
            when (check_all_text_view.text.toString()) {
                getString(R.string.check_all) -> adapter.setIsCheckedAllItem(true)
                getString(R.string.uncheck_all) -> adapter.setIsCheckedAllItem(false)
            }
        }
        delete_btn.setOnClickListener { deleteDialog.show(supportFragmentManager, null) }
        swipe_refresh_layout.setOnRefreshListener { refresh() }

        LocalBroadcastManager.getInstance(this)
            .registerReceiver(deleteReceiver, IntentFilter(ACTION_DELETE_ITEM))
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)

        swipe_refresh_layout.post {
            swipe_refresh_layout.isRefreshing = true
            refresh()
        }
    }

    private fun refresh() {
        viewModel.refresh()
            .`as`(DisposableConverter.dispose(this))
            .subscribe(object :
                RefreshObserver<PagedList<ContentEntity>>(swipe_refresh_layout) {
                override fun onNext(it: PagedList<ContentEntity>) {
                    super.onNext(it)
                    adapter.submitList(it)
                }

                override fun onError(e: Throwable) {
                    super.onError(e)
                    empty_view.setText(R.string.request_fail)
                }
            })
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> if (mIsDeleteEnabled) {
                setDeleteEnabled(false)
            } else {
                finish()
            }
        }
        return super.onOptionsItemSelected(item)
    }

    // 删除选中数据
    private fun deleteItems() {
        viewModel.deleteItems(adapter.deleteData())
            .`as`(DisposableConverter.dispose(this))
            // todo 写android.R.id.content行不行
            .subscribe(object : SnackbarObserver<Int>(swipe_refresh_layout) {
                override fun onNext(deleteCount: Int) {
                    super.onNext(deleteCount)
                    val text = String.format(getString(R.string.delete_successful, deleteCount))
                    Snackbar.make(swipe_refresh_layout, text, Snackbar.LENGTH_SHORT).show()
                    setDeleteEnabled(false)
                }
            })
    }

    // 设置是否可删除
    private fun setDeleteEnabled(isDeleteEnabled: Boolean) {
        this.mIsDeleteEnabled = isDeleteEnabled
        adapter.setDeleteBoxVisibility(isDeleteEnabled)
        val visibility: Int = if (isDeleteEnabled) {
            delete_btn.show()
            View.VISIBLE
        } else {
            delete_btn.hide()
            View.INVISIBLE
        }
        check_all_text_view.visibility = visibility
    }

    override fun onBackPressed() {
        if (mIsDeleteEnabled) {
            setDeleteEnabled(false)
        } else {
            super.onBackPressed()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        swipe_refresh_layout.setOnRefreshListener(null)
        LocalBroadcastManager.getInstance(this)
            .unregisterReceiver(deleteReceiver)
    }

    inner class DeleteReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
            val isDeleteAll = intent!!.getBooleanExtra(EXTRA_DELETE_ALL, false)
            check_all_text_view.text =
                getString(if (isDeleteAll) R.string.uncheck_all else R.string.check_all)
        }
    }

    companion object {
        const val ACTION_DELETE_ITEM = "action.DELETE_ITEM"
        const val EXTRA_DELETE_ALL = "EXTRA_DELETE_ALL"

        fun startActivity(context: Context) {
            val intent = Intent(context, CollectionActivity::class.java)
            context.startActivity(intent)
        }
    }
}